server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/ucacoin/UcacoinExtension.java

```diff
@@ -0,0 +1,135 @@
+import com.generalbytes.batm.server.extensions.*;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.wallets.UcacoinRPCWallet;
+import com.generalbytes.batm.server.extensions.extra.ucacoin.exchanges.bkex.BkexExchange;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.StringTokenizer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UcacoinExtension extends AbstractExtension{
+
+    private static final Logger LOG = LoggerFactory.getLogger("batm.master.bkexExchange");
+
+    @Override
+    public String getName() {
+        return "BATM Ucacoin extension";
+    }
+
+    @Override
+    public IWallet createWallet(String walletLogin, String tunnelPassword) {
+        if (walletLogin !=null && !walletLogin.trim().isEmpty()) {
+            StringTokenizer st = new StringTokenizer(walletLogin,":");
+            String walletType = st.nextToken();
+
+            if ("ucacoind".equalsIgnoreCase(walletType)) {
+                //"ucacoind:protocol:user:password:ip:port:accountname"
+
+                String protocol = st.nextToken();
+                String username = st.nextToken();
+                String password = st.nextToken();
+                String hostname = st.nextToken();
+                String port = st.nextToken();
+                String accountName ="";
+                if (st.hasMoreTokens()) {
+                    accountName = st.nextToken();
+                }
+
+
+                if (protocol != null && username != null && password != null && hostname !=null && port != null && accountName != null) {
+                    String rpcURL = protocol +"://" + username +":" + password + "@" + hostname +":" + port;
+                    return new UcacoinRPCWallet(rpcURL,accountName);
+                }
+            }
+            if ("ucademo".equalsIgnoreCase(walletType)) {
+
+                String fiatCurrency = st.nextToken();
+                String walletAddress = "";
+                if (st.hasMoreTokens()) {
+                    walletAddress = st.nextToken();
+                }
+
+                if (fiatCurrency != null && walletAddress != null) {
+                    return new DummyExchangeAndWalletAndSource(fiatCurrency, CryptoCurrency.UCA.getCode(), walletAddress);
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public ICryptoAddressValidator createAddressValidator(String cryptoCurrency) {
+        if (CryptoCurrency.UCA.getCode().equalsIgnoreCase(cryptoCurrency)) {
+            return new UcacoinAddressValidator();
+        }
+        return null;
+    }
+
+    @Override
+    public IRateSource createRateSource(String sourceLogin) {
+        if (sourceLogin != null && !sourceLogin.trim().isEmpty()) {
+            StringTokenizer st = new StringTokenizer(sourceLogin,":");
+            String exchangeType = st.nextToken();
+
+            if ("bkexRateSource".equalsIgnoreCase(exchangeType)) {
```

Comment:
make "bkexRateSource" lowercase = "bkexratesource"
don't forget to fix also xml.